Closes #2612 CPCSS for mobile: generation (3/4)

remyperona opened the pull request:
Closes #2612

### TODO

#### CriticalPath/AdminSubscriber
- [x] Unit tests failing for `enqueue_admin_edit_script()` => Tonya

#### inc/Engine/CriticalCSSSubscriber.php
- [x] Add new method `maybe_generate_cpcss_mobile()` to also trigger a generation when "Optimize CSS delivery" is enabled, "specific cache for mobile" is enabled and "async_css_mobile" = `1`
- [x] Update `init_critical_css_generation()` to change the CPCSS to generate depending on the options
- [x] Update `generate_critical_css_on_activation()` to change the CPCSS to generate depending on the options
- [x] Add unit & integration tests for `maybe_generate_cpcss_mobile()` => Ahmed
- [x] Add unit & integration tests for `init_critical_css_generation()` => Remy
- [x] Update unit & integration tests for `generate_critical_css_on_activation()` => Remy

#### inc/Engine/CriticalCSS.php
- [x] Update `set_items()` to also add mobile version of each item if the "specific cache for mobile" option is enabled and "async_css_mobile"  = `1`
- [x] Update `process_handler()` to set different items list depending on the type of CPCSS to generate
- [x] Add unit & integration tests for `process_handler()` => Ahmed
- [x] `clean_critical_css()` use an argument like we did for set_items(), to be able to do a partial clear only

#### inc/Engine/CriticalCSSGeneration.php
- [x] Update `task()`to use the new handshake with the API introduced in #2512. Re-queue the task when it's still in progress instead of using a `while` loop and `sleep`. Also make sure the `mobile` parameter is passed along the POST request when needed
- [x] Add unit & integration tests for `task()` => Remy

#### CPCSS for specific posts
- [x] Generate the mobile version when needed along the normal version
- [x] Delete the mobile version when needed along the normal version
- [x] Update tests => Ahmed and Cristina
- [x] Delete CPCSS for a post when the post is deleted => Cristina
- [x] See if we can display the notice on post differently depending on the result (if that’s not yet the case), by using the success/warning/error styles
Cristina - this was working just fine in Classic editor, but not in Block editor. Needed to add & remove some div class to make it work properly in Block Editor.
- [x] Don't generate CPCSS mobile for posts if the option "separate cache for mobile devices" is disabled => Ahmed

#### DataManager
- [x]  get_cache_job_id() - adds $is_mobile
- [x]  set_cache_job_id() - adds $is_mobile
- [x]  delete_cache_job_id() - adds $is_mobile
- [x] delete_cpcss() - adds $is_mobile 

#### RESTWPPOST
- [x] delete tests are failing => Tonya

engahmeds3ed committed:
add mobile parameter to the post request body and adjust tests based on that

remyperona committed:
Refactor the background task to use the ProcessorService

remyperona committed:
correctly add the path to each item

remyperona committed:
move docblock for filter in APIClient

remyperona committed:
Update to check for async_css_mobile and add mobile version

remyperona committed:
update path to use the filename

remyperona committed:
sort items in alphabetical order

engahmeds3ed committed:
adjust messages based on mobile

remyperona committed:
add mobile parameter to the process_generate request

remyperona committed:
fix unit tests

remyperona committed:
revert to public

remyperona committed:
fix phpcs

crystinutzaa committed:
Added Mobile CPCSS generation (WIP)

crystinutzaa committed:
Added missing param is_mobile

remyperona committed:
Update from parent

crystinutzaa committed:
Displayed 2 notices at CPCSS rest generation

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

Tabrisrp commented on views/metabox/cpcss/container.php:
```diff
@@ -28,3 +28,12 @@
+		</div>
+	</div>
+<?php endif; ?>
+<script type="text/javascript">
+	let wprMobileCpcssEnabled = <?php echo esc_js( $data['async_css_mobile'] ); ?>;
+</script>
```
This should be added in the `wp_localize_script()` in `AdminSubscriber::admin_enqueue_script()` instead

crystinutzaa replied:
Thank you! I was not aware about this till now. Modified 💯

engahmeds3ed committed:
adjust unit tests after changing the structure of RESTWP

engahmeds3ed committed:
refining generate route unit test

engahmeds3ed committed:
adjust cpcss section integration test

crystinutzaa committed:
Removed mobile notice

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

remyperona committed:
set mobile value correctly

remyperona committed:
generate mobile CPCSS on mobile cache activation

remyperona committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

remyperona committed:
Update from parent

crystinutzaa committed:
Calling mobile CPCSS deletion on Rest delete API

crystinutzaa committed:
Moved wprMobileCpcssEnabled js in wp_localize_script()

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile' into feature/cpcss-mobile-generation-3

crystinutzaa committed:
First round of fixing Unit Tests for CPCSS for specific posts

crystinutzaa committed:
Fixed PHPCS

engahmeds3ed committed:
add mobile tests to APIClient get_job_details method

engahmeds3ed committed:
Merge remote-tracking branch 'origin/feature/cpcss-mobile-generation-3' into feature/cpcss-mobile-generation-3

engahmeds3ed committed:
add mobile unit tests to APIClient send_generation_request method

engahmeds3ed committed:
add mobile integration tests to APIClient send_generation_request method

engahmeds3ed committed:
add mobile tests for DataManager

engahmeds3ed committed:
adjust mobile processorService => process_generate method

hellofromtonya committed:
Fixes CriticalPath\AdminSubscriber::enqueue_admin_edit_script() unit tests.

crystinutzaa committed:
Revisited enqueue_admin_edit_script() Tests to include async_css_mobile

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

crystinutzaa committed:
Fixed integration test for cpcss_section()

engahmeds3ed committed:
add unit tests for maybe_generate_cpcss_mobile

crystinutzaa committed:
Modified Unit & Integration tests from DataManager to include is_mobile

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

crystinutzaa committed:
Added is_mobile to ProcessorService::process_generate Unit test

engahmeds3ed committed:
add integration tests for maybe_generate_cpcss_mobile

hellofromtonya committed:
Fixes RESTWPPost::delete unit tests.

hellofromtonya committed:
Formatting.

remyperona committed:
add permissions check in init_critical_css_generation

remyperona committed:
add tests for init_critical_css_generation

remyperona committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

hellofromtonya committed:
Updates the RESTWPPost integration test DocBlock.

remyperona committed:
Update from parent

remyperona committed:
fix unit tests

remyperona committed:
update mock

hellofromtonya committed:
Adds mobile to RESTWPPost::delete() integration tests.

remyperona committed:
fix index check

remyperona committed:
update tests

hellofromtonya committed:
Adds DocBlocks.

hellofromtonya committed:
Updates CriticalCSS::set_items() @param description.

hellofromtonya committed:
Makes $version param description consistent.

remyperona committed:
update tests

remyperona committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

hellofromtonya committed:
Moves try/catch FilesystemIterator to private methods.

hellofromtonya committed:
Fixes variable name boo boo.

hellofromtonya committed:
Moves filesystem bailout checker to a private function to encapsulate the check.

remyperona committed:
fix test by deleting the file earlier

engahmeds3ed committed:
start unit tests for process_handler

hellofromtonya committed:
Injects filesystem handler into CriticalCSSSubscriber.

crystinutzaa committed:
Added Mobile CPCSS for Rest::generate()

hellofromtonya committed:
Updates CriticalCSSSubscriber for injecting filesystem + a few improvements.

hellofromtonya committed:
Adds @uses for test coverage.

remyperona committed:
add unit tests for task method

crystinutzaa committed:
Added Mobile CPCSS process generation from Tools view

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

remyperona committed:
start integration tests for task()

engahmeds3ed committed:
finish unit test for process_handler, needs some optimizations

engahmeds3ed committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

hellofromtonya committed:
Fixes the removal of rocket_get_constant() from CriticalCSS constructor. Whoopsie.

hellofromtonya committed:
Moves content setup and go_to to a trait for reuse.

hellofromtonya committed:
Improves get css content integration tests.

engahmeds3ed committed:
finish unit test for process_handler optimizations

Commit:
Fixes property name in integration test.

engahmeds3ed committed:
integration test for process_handler method

crystinutzaa committed:
Prevent Unit test from running (identify error)

crystinutzaa committed:
Bailout unit test

crystinutzaa committed:
Fixed namespace for test_Task Integration

crystinutzaa committed:
Mocking ProcessorService for task Integration test

hellofromtonya committed:
Improves CriticalCSS::process_handler() unit tests.

hellofromtonya committed:
Optimization: return from get_results() instead of assigning to var and then returning var.

hellofromtonya committed:
Upgrades CriticalCss::process_handler integration tests.

hellofromtonya committed:
Adds @uses to CriticalCSS::process_handler integration tests.

remyperona committed:
update variable name

remyperona committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

remyperona committed:
rename variable

Tabrisrp commented on inc/Engine/CriticalPath/RESTWP.php:
```diff
@@ -109,10 +136,13 @@ public function generate( WP_REST_Request $request ) {
-		$item_path = $this->get_path( $item_id );
 		$timeout   = ( isset( $request['timeout'] ) && ! empty( $request['timeout'] ) );
+		$item_path = $this->get_path( $item_id );
+		if ( $is_mobile ) {
+			$item_path = $this->get_path( $item_id, true );
+		}
```
```suggestion
		$item_path = $this->get_path( $item_id, $is_mobile );
```

hellofromtonya commented on inc/Engine/CriticalPath/CriticalCSS.php:
```diff
@@ -269,51 +277,75 @@ public function get_public_taxonomies() {
 		 */
 		$this->items = (array) apply_filters( 'rocket_cpcss_items', $this->items );
+
+		ksort( $this->items );
```
@Tabrisrp `ksort()` is slow. Why do we need this?

Tabrisrp replied:
The goal is to re-order the array to group the 2 versions (default & mobile) for the same page together, instead of doing all the default ones first, then all the mobile ones after

hellofromtonya replied:
Then shouldn't we only do it if we're also doing mobile, i.e. version === 'all'?

Tabrisrp replied:
you're right it's not needed for the other cases. Commit: https://github.com/wp-media/wp-rocket/commit/d04a9865962683ff3a85a390e457fba1a066775c

crystinutzaa committed:
Fixed get_path() call from generate

hellofromtonya commented on inc/Engine/CriticalPath/CriticalCSSGeneration.php:
```diff
@@ -49,141 +54,36 @@ class CriticalCSSGeneration extends WP_Background_Process {
-
+		$timeout   = false;
 		$transient = get_transient( 'rocket_critical_css_generation_process_running' );
+		$mobile    = isset( $item['mobile'] ) ? true : false;
```
`isset` returns `true and `false. We don't need to explicitly return `true` or `false`.

hellofromtonya commented on inc/Engine/CriticalPath/DataManager.php:
```diff
@@ -153,13 +178,14 @@ public function delete_cache_job_id( $item_url ) {
 	 */
-	private function get_cache_key_from_url( $item_url ) {
+	private function get_cache_key_from_url( $item_url, $is_mobile = false ) {
 		$encoded_url = md5( $item_url );
 
-		return 'rocket_specific_cpcss_job_' . $encoded_url;
+		return 'rocket_specific_cpcss_job_' . $encoded_url . ( $is_mobile ? '_mobile' : '' );
```
Optimization opportunity:

It's faster to not concatenate an empty string. 

```php
private function get_cache_key_from_url( $item_url, $is_mobile = false ) {
	$encoded_url = md5( $item_url );

	if ( $is_mobile ) {
		$encoded_url .= '_mobile';
	}

	return 'rocket_specific_cpcss_job_' . $encoded_url;
}
```

remyperona committed:
only sort when we need all types generated

hellofromtonya commented on inc/Engine/CriticalPath/ProcessorService.php:
```diff
@@ -156,14 +163,20 @@ private function check_cpcss_job_status( $job_id, $item_path, $item_url ) {
+			$error = sprintf( __( 'Mobile Critical CSS for %1$s not generated.', 'rocket' ), $item_url );
+		} else {
+			// translators: %1$s = page URL.
+			$error .= sprintf( __( 'Critical CSS for %1$s not generated.', 'rocket' ), $item_url );
```
Shouldn't this assign instead of concatenate and assign?

crystinutzaa committed:
Removed unused Is_mobile var

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

crystinutzaa committed:
Fixed changes requested by @hellofromtonya

Tabrisrp commented:
@wp-media/dev-qa Ready for QA.

Testing (considering that `async_css_mobile` is enabled):
- Enable "_optimize CSS delivery_" + "_separate cache for mobile_": default & mobile version of CPCSS should be generated
- "_Optimize CSS Delivery_" already enabled, enable "_separate cache for mobile_": mobile version of CPCSS should be generated
- Click on "_regenerate CPCSS_": appropriate version should be generated depending on the selected options (either only default, or default + mobile)
- On a post: click on generate, 2 files should be generated (default + mobile) if "_separate cache for mobile_" is enabled
- On a post: deleting the specific CPCSS should delete the 2 files

piotrbak reviewed (changes requested):
@Tabrisrp @hellofromtonya 

**The 2nd scenario failed:**
1. "Optimize CSS Delivery" already enabled
2. Enabling "Separate Cache for Mobile Devices"
3. Mobile Critical CSS is being generated
4. **Desktop Critical CSS is wiped out**

**Also, the 4th one is not working as expected:**
1. "Optimize CSS Delivery" enabled
2. "Separate Cache for Mobile Devices" disabled
3. Generating Critical CSS for post
4. Two versions of Critical CSS are generated

**Additional thoughts:**
1. Thrashing and removing post doesn't remove Critical CSS
2. We're showing the `yellow` popup for a success message. In most cases we use yellow when it's a warning:
![image](https://user-images.githubusercontent.com/26767037/82843176-3e146200-9ed4-11ea-81dc-d92d46c5e7a7.png)

3. The following scenario doesn't generate mobile Critical CSS:
- Enable "Optimize CSS Delivery" without "Separate Cache for Mobile Devices" enabled
- Files are generated, disable "Optimize CSS Delivery"
- Files are not removed
- Enable "Separate Cache for Mobile Devices"
- Enable "Optimize CSS Delivery"

4. Not sure if this can damage anything but:
- "Optimize CSS Delivery" enabled
- Open post editor
- Disable "Optimize CSS Delivery"
- We're still able to generate Critical CSS for a post correctly.

crystinutzaa commented:
@piotrbak where did you tested the color notice? Where I can see it. For this one:
` See if we can display the notice on post differently depending on the result (if that’s not yet the case), by using the success/warning/error styles` 
I am using these 2 classes for that notice: `notice-error` or `notice-success` depending on the reply from the REST API.

crystinutzaa committed:
Fixed notice display in Block Editor

crystinutzaa committed:
Delete CPCSS for a post when the post is deleted ( + fixed current unit & Integration tests)

hellofromtonya commented on tests/Unit/inc/Engine/CriticalPath/CriticalCSSSubscriber/maybeGenerateCpcssMobile.php:
```diff
@@ -33,7 +35,8 @@ public function setUp() {
 			null,
 		] );
-		$this->subscriber   = new CriticalCSSSubscriber( $this->critical_css, $options, null );
+		$this->processor_service = Mockery::mock( ProcessorService::class );
```
@crystinutzaa Are we mocking methods on the `ProcessorService`? If no, then we don't need to have this as a property on the test class.

engahmeds3ed replied:
I think this is resolved here: https://github.com/wp-media/wp-rocket/commit/3180e6571c892ccc5173782186766c1c0ea44b2e

remyperona committed:
remove files exist check

remyperona committed:
remove files exist test

remyperona committed:
remove files exist tests

remyperona committed:
update clean_critical_css to add version parameter

remyperona committed:
update clean_critical_css tests

engahmeds3ed committed:
adjust key value in term teardown

engahmeds3ed committed:
Don't generate CPCSS mobile for posts if the option "separate cache for mobile devices" is enabled and adjust tests

engahmeds3ed committed:
adjust phpcs format

crystinutzaa committed:
Added Integration & Unit tests for delete_cpcss()

crystinutzaa committed:
fixed fixture change

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

crystinutzaa committed:
removed property from maybeGenerateCpccMobile

Tabrisrp commented on inc/Engine/CriticalPath/RESTWP.php:
```diff
@@ -100,7 +110,39 @@ public function check_permissions() {
+			);
+		}
+
+		// Bailout in case separate cache for mobile option is enabled.
+		if ( $is_mobile && $this->options->get( 'do_caching_mobile_files', 0 ) ) {
+			return rest_ensure_response(
+				$this->return_error(
+					new WP_Error(
+						'caching_mobile_separate_enabled',
+						__( 'Separate cache files for mobile devices option is enabled.', 'rocket' ),
+						[
+							'status' => 400,
+						]
+					)
+				)
+			);
+		}
```
I don't understand this part, is there something missing?

hellofromtonya replied:
@Tabrisrp It's part of this commit https://github.com/wp-media/wp-rocket/commit/3a7adbfc5607f65455d91d981465a9fcfc94dbd6. What should this be?

Tabrisrp replied:
it should be the opposite, we should prevent generation if the option is disabled

Tabrisrp replied:
https://github.com/wp-media/wp-rocket/pull/2672/commits/726e28ac72a6d497cfec4e09e0f98e9059a5e939

hellofromtonya committed:
Updates to formatting and param comments.

remyperona committed:
Update bailout for separate mobile cache

Tabrisrp commented:
@piotrbak This is ready for a new QA.

- Scenario 2 & 4 should be fixed
- Deleting a post should also delete the related CPCSS (still preserved when trashing, in case the post is restored)
- Notices display is updated
- Mobile critical CSS should be correctly generated for your point 3
- No changes for point 4 as it's pretty edge case and doesn't break anything

piotrbak commented:
Starting the QA

piotrbak commented:
@wp-media/dev 
I have problems with the last scenario (testing on Cloudflare), I can't tell you if it's related to the Cloudflare itself or the changes made after my last tests. When trying to switch the post to the regular CPCSS, I'm getting 403 error:
![403](https://user-images.githubusercontent.com/26767037/82959502-c582d380-9faf-11ea-9d6e-068d4e7fa140.PNG)
I'm not able to test it in the same environment as yesterday since something is broken on Mega now.

My second concern is that I'm very rarely able to generate CPCSS for all post types. On my personal website, where I have 24 to generate, I wasn't able to do it even once. After some requests we're receiving:
`timeout. Please retry a little later.`
Maybe we can somehow slow down the process, in order not to make requests that fast? I'm not sure if this is the real culprit though.

@arunbasillal Could you test those things on your own?
On sites with and without Cloudflare:
- Generate CPCSS for a single post
- Try to revert it back to the regular CPCSS
- Is there a 403?

On the same sites, check if you're able to generate CPCSS for all post types without any problems (give it a couple of tries). Thanks!

crystinutzaa commented:
@piotrbak the 403 on Revert back to regular CPCSS is related to a security plugin (probably Cloudflare) which blocks the DELETE REST request.
As for generation, the request is sent only once & then it simply rechecks the result from the API. 
It can be something in the CPCSS API which causes this.

@Tabrisrp who can check why this happens?

Commit:
CPCSS admin classes architecture update (PR #2687)

* Add Post & Settings classes
* Move & rename subscriber, update dependencies
* update classes instantiation
* update all tests

Commit:
CPCSS for mobile: pending generation handler (PR #2686)

hellofromtonya commented:
@piotrbak This PR is ready for re-QA. 

What changed? You found a time-out issue. That was being caused by us blasting the API for requests to check if it was done processing or not. We weren't waiting to give it time to get done.

In PR 2686, we added an async system that waits 3 seconds to ask the API if it's done.

Please redo your QA on this PR. Thanks!

arunbasillal commented:
Doing the QA again. Noticed this: 

![cpcss](https://user-images.githubusercontent.com/5751474/83530312-2dce3980-a509-11ea-9bd6-3ad72bbdb332.png)

Some early feedback: 

- The mobile version is currently counted as a post type. This shouldn't be the case because mobile version isn't a post type. 
- The count seems to be wrong as you can see in the screenshot. We had something similar in the past, maybe it's related? - https://github.com/wp-media/wp-rocket/issues/1333
- We had agreed not to display separate notices for mobile, sorry if this was not communicated widely - https://wp-media.slack.com/archives/G8L4EE8N9/p1590096487031800?thread_ts=1590088503.020600&cid=G8L4EE8N9
- ^ separate notice for mobile can be confusing for customers. They will wonder why they are not seeing it (i.e. when separate cache for mobile is not enabled). 

Continuing with testing.

hellofromtonya commented:
@crystinutzaa See @arunbasillal's notes above about not having a mobile separate notices. Thoughts?

arunbasillal commented:
- Enable "optimize CSS delivery" + "separate cache for mobile": default & mobile version of CPCSS should be generated 👍 
- ^ Tested if the mobile version is actually used on mobile. 👍 
- ^ When mobile version was not available, desktop CPCSS was used as fallback 👍 

- "Optimize CSS Delivery" already enabled, enable "separate cache for mobile": mobile version of CPCSS should be generated 👍 
- ^Desktop CPCSS is not wiped out 👍 

- Click on "regenerate CPCSS": appropriate version should be generated depending on the selected options (either only default, or default + mobile) 👍 

- On a post: click on generate, 2 files should be generated (default + mobile) if "separate cache for mobile" is enabled 👍 
- ^ Tested if the mobile version is actually used on mobile. 👍 
- ^ When mobile version was not available, desktop CPCSS was used as fallback 👍 
- ^ the generated notice is green:
![image](https://user-images.githubusercontent.com/5751474/83535472-99b3a080-a50f-11ea-9fed-feabf2ea6a06.png)
- ^ Tested with classic editor as well 👍 
- ^ When "Separate cache for mobile is disabled", only desktop version of specific CPCSS is generated. 👍 

- On a post: deleting the specific CPCSS should delete the 2 files 👍 
- ^ The notice is green
![image](https://user-images.githubusercontent.com/5751474/83535537-acc67080-a50f-11ea-8114-d5fea048fb07.png)
- ^ Tested with classic editor as well 👍 
- ^ Trashing a post deletes it CPCSS as well 👍 

Everything good so far 🚀 

**NOTES**

On the [mega site](http://mega.wp-rocket.me/wp-admin/options-general.php?page=wprocket#cache), I am seeing the case of the infinite loop (#1333): 
![image](https://user-images.githubusercontent.com/5751474/83534196-2d846d00-a50e-11ea-81be-2c05655b9363.png)

Something else regarding the position of the "Regenerate Critical Path CSS" item in the admin bar. 

- On this feature branch the admin bar item is the last one after "Support"
![image](https://user-images.githubusercontent.com/5751474/83541395-375e9e00-a517-11ea-979a-6961c06bf435.png)

- This is how it used to be, even on version 3.6-alpha1:
![image](https://user-images.githubusercontent.com/5751474/83541540-637a1f00-a517-11ea-88c0-faa96267e436.png)

- Why was the position moved? 

Tested this scenario that Piotr recorded: 
- Enable "Optimize CSS Delivery" without "Separate Cache for Mobile Devices" enabled
- Files are generated, disable "Optimize CSS Delivery"
- Files are not removed
- Enable "Separate Cache for Mobile Devices"
- Enable "Optimize CSS Delivery"

Result: CPCSS for mobile was deleted correctly. 
But the desktop CPCSS was deleted first and then regenerated. Is this necessary? Video: https://youtu.be/k1dyjQ3_TTs

hellofromtonya commented:
@crystinutzaa @Tabrisrp Please look at the Arun's notes and questions including:

- On the mega site, I am seeing the case of the infinite loop (#1333)
- Something else regarding the position of the "Regenerate Critical Path CSS" item in the admin bar.
- But the desktop CPCSS was deleted first and then regenerated. Is this necessary?
- Mobile notices

Are there actions we need to take?

Tabrisrp commented:
> On the mega site, I am seeing the case of the infinite loop (#1333)

Was it infinite, or did it stop eventually? I'm wondering if it's related to what we discussed last week, with the possible read/write of the transients at the same time by 2 different processes.

> Something else regarding the position of the "Regenerate Critical Path CSS" item in the admin bar.

This is because of the change I made to bring this admin bar item code into the module. The admin bar system doesn't contain a way to order sub-items like this, except by using hook priority. To preserve the existing ordering, I would have to re-do all the related admin bar code in the plugin to use appropriate priorities. This didn't fit into what we are doing here, so I went with the current approach, that is not perfect but preserve the functionality.

> But the desktop CPCSS was deleted first and then regenerated. Is this necessary?

This scenario considers that we want to generate both versions of the files, because "Optimize CSS Delivery" was not enabled before (even though the desktop files exists). So it's going to delete everything, and recreate them. We also have no way to know if the existing files are still valid or not, maybe they're old, or they changed plugins/themes in between changing the "Optimize CSS Delivery" option value?

> Mobile notices

We can skip adding them into the notice if it's the approach that the product team wants. It seems to be the original intent, but it was missed in our communication until now. Are we fine with users not knowing if a mobile version was not generated correctly?

hellofromtonya commented:
>Was it infinite, or did it stop eventually? I'm wondering if it's related to what we discussed last week, with the possible read/write of the transients at the same time by 2 different processes.

How can we validate the _**root cause**_ of this problem? @Tabrisrp

hellofromtonya commented:
For Mobile Notices:

>We can skip adding them into the notice if it's the approach that the product team wants. It seems to be the original intent, but it was missed in our communication until now. Are we fine with users not knowing if a mobile version was not generated correctly?

Yes, this is something Product wants:

>>We had agreed not to display separate notices for mobile, sorry if this was not communicated widely - https://wp-media.slack.com/archives/G8L4EE8N9/p1590096487031800?thread_ts=1590088503.020600&cid=G8L4EE8N9

Let's create a PR to remove the mobile notices and update the tests please. @crystinutzaa @Tabrisrp

Tabrisrp commented:
> Does this mean that eventually as we redo other aspects of the admin bar, we will have the original order in the future?

Yes, in the end we will

arunbasillal commented:
@Tabrisrp I had a chat with Jonathan regarding the position of "Regenerate CPCSS" button. Adding it in the end doesn't look good and can be confusing for customers. 

Jonathan said, `A code update shouldn’t change the current state / behavior of the plugin`. 

Shall we do one of these?
- Update everything in the admin bar so that the original order is preserved. Not sure how much time it needs. 
- Revert it to how it used to be. We can update everything related to this in a different sprint. 

cc: @hellofromtonya

Commit:
Update WP Rocket admin bar items order (PR #2701)

* add new action hook in the admin bar function
* update hook name to hook on the new admin bar action

crystinutzaa committed:
Removed transient generate + added URL success or failure

Commit:
Remove CPCSS mobile versions from admin notice (#2700)

Don't display the status of the CPCSS mobile version generation in the admin notice

hellofromtonya committed:
Minor refactor for setting "success" only if code is cpcss_generation_successful.

crystinutzaa committed:
Fixed PHPCS

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

hellofromtonya committed:
Removes delete_transient() from generation_complete. Why? No need to delete it again.

crystinutzaa committed:
Removed generated from Unit tests ( Round 1)

crystinutzaa committed:
Merge branch 'feature/cpcss-mobile-generation-3' of https://github.com/wp-media/wp-rocket into feature/cpcss-mobile-generation-3

crystinutzaa committed:
Finished Unit tests without generated

crystinutzaa committed:
Removed generated from Integration tests

hellofromtonya committed:
Fixes Admin\Admin::cpcss_heartbeat unit tests.

hellofromtonya committed:
A few updates to the test fixture and unit test.

hellofromtonya commented:
@arunbasillal @piotrbak This PR is ready for re-QA.

Things fixed from Arun's last QA:

- Only disables desktop notices (no mobile notices)
- Fixed the order of the admin bar menu items
- Fixed the count

arunbasillal commented:
- Notices doesn't show separate mobile CPCSS status 👍 
- Admin bar looks good 👍 
- Count looks good 👍 

Noticed an issue, to replicate:

- Make sure "Separate Cache for mobile" is disabled. 
- Enable Optimize CSS delivery. Let the generation complete. 
- Enable "Separate cache for mobile"
- This error happens. But file generated correctly. 
![image](https://user-images.githubusercontent.com/5751474/83600597-87c41300-a58c-11ea-9f9e-b7509855528b.png)
Video: https://youtu.be/fyfMEQv79x0

Another issue:

- When separate cache for mobile is enabled
- The complete notice says "complete" but in `blue` and asks customers to refresh 
![image](https://user-images.githubusercontent.com/5751474/83601168-ad9de780-a58d-11ea-8821-6f9e262bcfa1.png)
- Then it completes with the same message
![image](https://user-images.githubusercontent.com/5751474/83601219-c7d7c580-a58d-11ea-9b20-3311192b4ffa.png)

Maybe this is the time it takes to generate mobile CPCSS? Can this situation be avoided? 
Video (from 1:25s - https://youtu.be/qLwKGPwjSzI)

crystinutzaa committed:
Fixed unwanted CPCSS notice - bailout if total is 0 & success is 0 (means we are in mobile mode only)

crystinutzaa commented:
@arunbasillal I fixed the first issue with the "unwanted 0 notice". Can you please recheck and let me know if is ok.

As for the second issue, the issue is with Mobile CPCSS. It takes time to complete so the notice will be displayed only when it finished totally. 
This is mandatory because we have 2 major actions on it:
```
                 /**
		 * Fires when the critical CSS generation process is complete.
		 *
		 * @since 2.11
		 */
		do_action( 'rocket_critical_css_generation_process_complete' );

		rocket_clean_domain();
```
Basically if we display the complete notice faster ( and the mobile did not finished) then rocket_clean_domain() will not run until the mobile is finished. So the customer will simply report that something if off if it does not wait / recheck in a while after the mobile generation finished too.
I don't think this notice should display the complete until mobile is generated too. Otherwise it may lead to a lot of confusion. 

@Tabrisrp & @hellofromtonya & @GeekPress ideas?

For that notice we can simply it and provide the following format. Basically the notice will contain both notices for Mobile & Desktop but with the details hidden. And we can provide a link / button (details) when clicked will enlarge and display the full list with each url. In this way the customer can be 100% informed about the CPCSS status.

https://www.screencast.com/t/2j7iIhHUc9Z

hellofromtonya commented:
The blue and then green messaging is due to non-mobile is done but mobile is still generating. Therefore, **_CPCSS generation is not done._**

CPCSS generation is done when both mobile and non-mobile are done for each path/URL.

Given that:

@crystinutzaa We'd need to change the behavior of the notice handler to only update the notice and count when both mobile and non-mobile are done for each path/URL.

By doing 👆, then the message will stay in sync and color match for the user's experience. The updating of the notice will be slower because we have 2 tasks running for each URL/path. But that's to be expected.

hellofromtonya commented:
@arunbasillal @GeekPress We are moving forward with [what I proposed above](https://github.com/wp-media/wp-rocket/pull/2672#issuecomment-638082701): 

- Notice and count for each URL/path only display _when_ both non-mobile and mobile CPCSS generation are done.

The team is brainstorming. We'll keep you in the loop.

arunbasillal commented:
- I retested the `unwanted 0 notice` and it is gone 👍 
- Mobile CPCSS is generated in the background 👍 

Thanks for the explanation @hellofromtonya. This is the ideal behavior. 

@crystinutzaa To answer your suggestions, the idea of not showing mobile specific CPCSS related notice is not to confuse customers. We generate it only when `separate cache for mobile` is enabled and having it on some websites and not having it on others (or when they change the `separate cache for mobile` setting) can be confusing. 

- The probability of desktop CPCSS generation being successful and mobile CPCSS generation to fail is low (this is an assumption of course). 
- Even if mobile CPCSS generation fails, the desktop CPCSS will be used as fallback. From our experience so far, for most responsive themes this is perfectly fine.

crystinutzaa committed:
Fixed CPCSS notice display

crystinutzaa committed:
Fixed fixture to accomodate new notice transient array format

arunbasillal reviewed (approved):
Did a quick retest:

- Separate notice for mobile CPCSS is not displayed 👍 
- "unwanted 0 notice" stays fixed 👍 
- Count of post types and generated post types is consistent (no infinite or excess count) 👍 
- Notice and count for each URL/path only display when both non-mobile and mobile CPCSS generation are done. 👍 

Everything reported in the last two days looks good so far. I would love to repeat the QA completely from the beginning, but it's a long one and its too late for me today. I will do it on the alpha version. 

Thanks and congrats to everyone 🚀
